File: apps/web/src/features/users/lib/referral-evidence.test.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
import assert from 'node:assert/strict'
import { test } from 'node:test'

import { inspectReferralEvidence } from './referral-evidence.ts'
import { createReferralSubmission } from './referral-submission.ts'

for (const character of ['a', '证', 'é', '🙂', '𠀀']) {
  test(`evidence counts Unicode characters: ${character}`, () => {
    for (const count of [1, 250, 334, 999, 1000, 1001]) {
      const value = character.repeat(count)
      assert.deepEqual(inspectReferralEvidence(value), {
        value,
        length: count,
        valid: count <= 1000,
      })
    }
  })
}

test('normalization preserves evidence and counts combining code points', () => {
  const padded = ' \u0085\u00a0证据\u3000 '
  assert.equal(inspectReferralEvidence(padded).value, '证据')
  assert.equal(
    inspectReferralEvidence('\ufeffevidence').value,
    '\ufeffevidence'
  )
  assert.equal(inspectReferralEvidence('e\u0301'.repeat(500)).valid, true)
  assert.equal(inspectReferralEvidence('e\u0301'.repeat(501)).valid, false)
  assert.equal(
    inspectReferralEvidence('  已核实\n保留原始记录  ').value,
    '已核实\n保留原始记录'
  )
})

test('empty, NUL and lone surrogate evidence is rejected before submission', () => {
  for (const value of [
    '',
    ' \t\r\n',
    '\u0085\u00a0\u3000',
    'before\0after',
    '\ud800',
    '\udfff',
  ]) {
    assert.equal(inspectReferralEvidence(value).valid, false)
  }
})

test('invalid input cannot reserve a request; retries retain exact text', () => {
  const submission = createReferralSubmission<{
    request_id: string
    evidence: string
  }>()
  let identities = 0
  const submit = (value: string) => {
    const input = inspectReferralEvidence(value)
    if (!input.valid) return undefined
    return submission(() => ({
      request_id: `test-request-${++identities}`,
      evidence: input.value,
    }))
  }
  assert.equal(submit('证'.repeat(1001)), undefined)
  assert.equal(identities, 0)
  const accepted = submit('  ' + '证'.repeat(1000) + '  ')
  assert.equal(accepted?.evidence, '证'.repeat(1000))
  assert.equal(identities, 1)
  // A failed response must not cause the next click to send a new operation.
  assert.equal(submit('another attempt'), accepted)
  assert.equal(identities, 1)
  assert.equal(Object.isFrozen(accepted), true)
})

File: apps/api-go/model/referral_evidence_test.go
package model

import (
	"strings"
	"testing"
)

func TestReferralEvidenceUnicodeBoundaries(t *testing.T) {
	for _, text := range []string{"a", "证", "é", "🙂", "\U00020000"} {
		for _, count := range []int{1, 250, 334, 999, 1000, 1001} {
			value := strings.Repeat(text, count)
			got, valid := normalizeReferralEvidence(value)
			if got != value || valid != (count <= 1000) {
				t.Fatalf("rune=%q count=%d bytes=%d valid=%v", text, count, len(value), valid)
			}
		}
	}
}

func TestReferralEvidenceNormalizesWithoutChangingAuditContent(t *testing.T) {
	for _, value := range []string{"  已核实\n保留原始记录  ", "\u0085\u00a0证据\u3000", "e\u0301", "\ufeffevidence"} {
		got, valid := normalizeReferralEvidence(value)
		if !valid || got != strings.TrimSpace(value) {
			t.Fatalf("normalization changed content: %q => %q", value, got)
		}
	}
	// A combining sequence is two code points, as it is in PostgreSQL.
	_, valid := normalizeReferralEvidence(strings.Repeat("e\u0301", 500))
	if !valid {
		t.Fatal("1000 code points rejected")
	}
	_, valid = normalizeReferralEvidence(strings.Repeat("e\u0301", 501))
	if valid {
		t.Fatal("1002 code points accepted")
	}
}

func TestReferralEvidenceRejectsEmptyNULAndInvalidUTF8(t *testing.T) {
	for _, value := range []string{"", " \t\r\n", "\u0085\u00a0\u3000", "before\x00after", string([]byte{0xff}), "bad\xc0\xaf"} {
		if _, valid := normalizeReferralEvidence(value); valid {
			t.Fatalf("invalid evidence accepted: %q", value)
		}
	}
}
